compiler/qsc_qasm3/src/lex/cooked.rs

```diff
@@ -0,0 +1,774 @@
+//! and `or`, which are cooked into [`BinaryOperator`] so that `and=` and `or=` are lexed correctly.
+//!
+//! Whitespace and comment tokens are discarded; this means that cooked tokens are not necessarily
+//! contiguous, so they include both a starting and ending byte offset.
+//!
+//! Tokens never contain substrings from the original input, but are simply labels that refer back
+//! to regions in the input. Lexing never fails, but may produce error tokens.
+
+#[cfg(test)]
+mod tests;
+
+use super::{
+    raw::{self, Number, Single},
+    Delim, Radix,
+};
+use crate::keyword::Keyword;
+use enum_iterator::Sequence;
+use miette::Diagnostic;
+use qsc_data_structures::span::Span;
+use std::{
+    fmt::{self, Display, Formatter},
+    iter::Peekable,
+    str::FromStr,
+};
+use thiserror::Error;
+
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub(crate) struct Token {
+    pub(crate) kind: TokenKind,
+    pub(crate) span: Span,
+}
+
+#[derive(Clone, Copy, Debug, Diagnostic, Eq, Error, PartialEq)]
+pub enum Error {
+    #[error("expected {0} to complete {1}, found {2}")]
+    #[diagnostic(code("Qasm3.Lex.Incomplete"))]
+    Incomplete(raw::TokenKind, TokenKind, raw::TokenKind, #[label] Span),
+
+    #[error("expected {0} to complete {1}, found EOF")]
+    #[diagnostic(code("Qasm3.Lex.IncompleteEof"))]
+    IncompleteEof(raw::TokenKind, TokenKind, #[label] Span),
+
+    #[error("unterminated string literal")]
+    #[diagnostic(code("Qasm3.Lex.UnterminatedString"))]
+    UnterminatedString(#[label] Span),
+
+    #[error("string literal with an invalid escape sequence")]
+    #[diagnostic(code("Qasm3.Lex.InvalidEscapeSequence"))]
+    InvalidEscapeSequence(#[label] Span),
+
+    #[error("unrecognized character `{0}`")]
+    #[diagnostic(code("Qasm3.Lex.UnknownChar"))]
+    Unknown(char, #[label] Span),
+}
+
+impl Error {
+    pub(crate) fn with_offset(self, offset: u32) -> Self {
+        match self {
+            Self::Incomplete(expected, token, actual, span) => {
+                Self::Incomplete(expected, token, actual, span + offset)
+            }
+            Self::IncompleteEof(expected, token, span) => {
+                Self::IncompleteEof(expected, token, span + offset)
+            }
+            Self::UnterminatedString(span) => Self::UnterminatedString(span + offset),
+            Self::InvalidEscapeSequence(span) => Self::InvalidEscapeSequence(span + offset),
+            Self::Unknown(c, span) => Self::Unknown(c, span + offset),
+        }
+    }
+
+    pub(crate) fn span(self) -> Span {
+        match self {
+            Error::Incomplete(_, _, _, s)
+            | Error::IncompleteEof(_, _, s)
+            | Error::UnterminatedString(s)
+            | Error::InvalidEscapeSequence(s)
+            | Error::Unknown(_, s) => s,
+        }
+    }
+}
+
+/// A token kind.
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Sequence)]
+pub enum TokenKind {
+    Annotation,
+    Pragma,
```

Comment:
Doc comments for these would be good showing what they look like in source code (just as other `TokenKind`s have)